src/librustc_mir/borrow_check/nll/region_infer/mod.rs

```diff
@@ -1096,6 +1136,32 @@ impl<'tcx> RegionInferenceContext<'tcx> {
         }).is_some()
     }
 
+    /// This function classifies a constraint from a location.
+    fn classify_constraint(&self, location: Location, mir: &Mir<'tcx>) -> ConstraintCategory {
```

Comment:
as an aside, it feels pretty clear we should think about breaking this logic out into its own module at some point. Doesn't have to be in this PR.

Reply:
I'll make an `error_reporting` module and move relevant functions into it.